From 42ed35f8f9c76ff56afdda9b0d3add958936bea6 Mon Sep 17 00:00:00 2001
From: P33M <P33M@github.com>
Date: Mon, 5 Aug 2013 13:17:58 +0100
Subject: [PATCH 089/174] dwc_otg: prevent leaking URBs during enqueue

A dwc_otg_urb would get leaked if the HCD enqueue function
failed for any reason. Free the URB at the appropriate points.
---
 drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c | 8 ++++++++
 1 file changed, 8 insertions(+)

--- a/drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c
+++ b/drivers/usb/host/dwc_otg/dwc_otg_hcd_linux.c
@@ -797,11 +797,19 @@ static int dwc_otg_urb_enqueue(struct us
 #if USB_URB_EP_LINKING
 			usb_hcd_unlink_urb_from_ep(hcd, urb);
 #endif
+			DWC_FREE(dwc_otg_urb);
 			urb->hcpriv = NULL;
 			if (retval == -DWC_E_NO_DEVICE)
 				retval = -ENODEV;
 		}
 	}
+#if USB_URB_EP_LINKING
+	else
+	{
+		DWC_FREE(dwc_otg_urb);
+		urb->hcpriv = NULL;
+	}
+#endif
 	DWC_SPINUNLOCK_IRQRESTORE(dwc_otg_hcd->lock, irqflags);
 	return retval;
 }
